pkg/promtail/api/entry_parser_test.go

```diff
@@ -18,59 +20,79 @@ type Entry struct {
+	{TestTimeStr + " stderr P message1\nmessage2", false, NewEntry(TestTime, "message1\nmessage2", "stderr")},
 }
 
 func TestCRI(t *testing.T) {
-	for _, tc := range criTestCases {
+	runTestCases(CRI, criTestCases, t)
+}
+
+var dockerTestCases = []TestCase{
+	{
+		Line:     "{{\"log\":\"bad json, should fail to parse\\n\",\"stream\":\"stderr\",\"time\":\"2019-03-04T21:37:44.789508817Z\"}",
+		Error:    true,
+		Expected: Entry{},
+	},
+	{
+		Line:  "{\"log\":\"some silly log message\\n\",\"stream\":\"stderr\",\"time\":\"2019-03-04T21:37:44.789508817Z\"}",
+		Error: false,
+		Expected: NewEntry(time.Date(2019, 03, 04, 21, 37, 44, 789508817, time.UTC),
+			"some silly log message\n",
+			"stderr"),
+	},
+	{
+		Line:  "{\"log\":\"10.15.0.5 - - [04/Mar/2019:21:37:44 +0000] \\\"POST /api/prom/push HTTP/1.1\\\" 200 0 \\\"\\\" \\\"Go-http-client/1.1\\\"\\n\",\"stream\":\"stdout\",\"time\":\"2019-03-04T21:37:44.790195228Z\"}",
+		Error: false,
+		Expected: NewEntry(time.Date(2019, 03, 04, 21, 37, 44, 790195228, time.UTC),
+			"10.15.0.5 - - [04/Mar/2019:21:37:44 +0000] \"POST /api/prom/push HTTP/1.1\" 200 0 \"\" \"Go-http-client/1.1\"\n",
+			"stdout"),
+	},
+}
+
+func TestDocker(t *testing.T) {
+	runTestCases(Docker, dockerTestCases, t)
+}
+
+func runTestCases(parser EntryParser, testCases []TestCase, t *testing.T) {
+	for i, tc := range testCases {
 		client := &TestClient{
 			Entries: make([]Entry, 0),
 		}
 
-		EntryParser := CRI.Wrap(client)
-		err := EntryParser.Handle(model.LabelSet{}, time.Now(), tc.Line)
+		handler := parser.Wrap(client)
+		err := handler.Handle(model.LabelSet{}, time.Now(), tc.Line)
 		hasError := err != nil
 
 		if tc.Error != hasError {
-			t.Error("For", tc.Line, "expected", tc.Error, "got", hasError)
+			t.Fatal("Parse error for test case", i, "with entry", tc.Line, "\nerror:", err)
 		}
 
 		if !tc.Error {
-			if len(client.Entries) != 1 {
-				t.Error("Handler did not receive the correct number of Entries, expected 1 received", len(client.Entries))
-			}
-
+			require.Equal(t, 1, len(client.Entries), "Handler did not receive the correct number of Entries")
 			entry := client.Entries[0]
-
-			if tc.Expected.Time != entry.Time {
-				t.Error("For", tc.Line, "expected", tc.Expected.Time, "got", entry.Time)
-			}
-
-			if tc.Expected.Log != entry.Log {
-				t.Error("For", tc.Line, "expected", tc.Expected.Log, "got", entry.Log)
-			}
-
-			if !tc.Expected.Labels.Equal(entry.Labels) {
-				t.Error("For", tc.Line, "expected", tc.Expected.Labels, "got", entry.Labels)
-			}
+			assert.Equal(t, tc.Expected.Time, entry.Time, "Time error for test case %d, with entry %s", i, tc.Line)
+			assert.Equal(t, tc.Expected.Log, entry.Log, "Log entry error for test case %d, with entry %s", i, tc.Line)
+			assert.True(t, tc.Expected.Labels.Equal(entry.Labels), "Label error for test case %d, labels did not match; expected: %s, found %s", i, tc.Expected.Labels, entry.Labels)
```

Comment:
Why can't you do `assert.Equal(t, tc.Expected, entry)`?

Reply:
Could simplify the first two, but I think I prefer the idea of using the prometheus Labels.Equal function for making sure the labels match?